src/proxy-cssom.js

```diff
@@ -0,0 +1,211 @@
+      }
+    }
+    window.CSSKeywordValue = ProxyCSSKeywordValue;
+  }
+
+  function toCssUnitValue(v) {
+    if (typeof v === 'number')
+      return new CSSUnitValue(v, 'number');
+    return v;
+  }
+
+  function toCssNumericArray(values) {
+    const result = [];
+    for (let i = 0; i < values.length; i++) {
+      result[i] = toCssUnitValue(values[i]);
+    }
+    return result;
+  }
+
+  class ProxyMathOperation {
+    constructor(values, operator, opt_name, opt_delimiter) {
+      privateDetails.set(this, {
+        values: toCssNumericArray(values),
+        operator: operator,
+        name: opt_name || operator,
+        delimiter: opt_delimiter || ', '
+      });
+    }
+
+    get operator() {
+      return privateDetails.get(this).operator;
+    }
+
+    get values() {
+      return  privateDetails.get(this).values;
+    }
+
+    toString() {
+      const details = privateDetails.get(this);
+      return `${details.name}(${details.values.join(details.delimiter)})`;
+    }
+  }
+
+  if (!window.CSSMathSum) {
+    class ProxyCSSMathSum extends ProxyMathOperation  {
+      constructor(values) {
+        super(arguments, 'sum', 'calc', ' + ');
+      }
+    }
+    window.CSSMathSum = ProxyCSSMathSum;
+  }
+
+  if (!window.CSSMathProduct) {
+    class ProxyCSSMathProduct extends ProxyMathOperation  {
+      constructor(values) {
+        super(arguments, 'product', 'calc', ' * ');
+      }
+    }
+    window.CSSMathProduct = ProxyCSSMathProduct;
+  }
+
+  if (!window.CSSMathNegate) {
+    class ProxyCSSMathNegate extends ProxyMathOperation {
+      constructor(values) {
+        super([arguments[0]], 'negate', '-');
+      }
+    }
+    window.CSSMathNegate = ProxyCSSMathNegate;
+  }
+
+  if (!window.CSSMathInvert) {
+    class ProxyCSSMathNegate extends ProxyMathOperation {
+      constructor(values) {
+        super([1, arguments[0]], 'invert', 'calc', ' / ');
+      }
+    }
+    window.CSSMathNegate = ProxyCSSMathNegate;
+  }
+
+  if (!window.CSSMathMax) {
+    class ProxyCSSMathMax extends ProxyMathOperation {
+      constructor() {
+        super(arguments, 'max');
+      }
+    }
+    window.CSSMathMax = ProxyCSSMathMax;
+  }
+
+  if (!window.CSSMathMin) {
+    class ProxyCSSMathMin extends ProxyMathOperation  {
+      constructor() {
+        super(arguments, 'min');
+      }
+    }
+    window.CSSMathMin = ProxyCSSMathMin;
+  }
+
+  if (!window.CSS)
+    window.CSS = {};
+
+  [
+    'number',
+    'percent',
+    // Length units
+    'em',
+    'ex',
+    'px',
+    'cm',
+    'mm',
+    'in',
+    'pt',
+    'pc',  // Picas
+    'Q',  // Quarter millimeter
+    'vw',
+    'vh',
+    'vmin',
+    'vmax',
+    'rems',
+    "ch",
+    // Angle units
+    'deg',
+    'rad',
+    'grad',
+    'turn',
+    // Time units
+    'ms',
+    's',
+    'Hz',
+    'kHz',
+    // Resolution
+    'dppx',
+    'dpi',
+    'dpcm',
+    // Other units
+    "fr"
+  ].forEach((name) => {
+    if (!CSS[name]) {
```

Comment:
Rather than checking existence for, and optionally installing each type, should we just have all of this conditional on the existence of window.CSS or on the existence of CSSUnitValue? I suspect that if a browser didn't support 'dpcm' for example, but did support CSSUnitValue, that trying to construct a CSSUnitValue(value, 'dpcm') would just throw an error, and so exposing the type in the CSS object is potentially introducing errors through suggested support for units.

Reply:
Done